src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs

```diff
@@ -2,24 +2,149 @@
+
+        public static bool SupportsGetAddrInfoAsync { get; } = Interop.Sys.PlatformSupportsGetAddrInfoAsync();
 
         public static void EnsureSocketsAreInitialized() { } // No-op for Unix
 
-        internal static Task GetAddrInfoAsync(string hostName, bool justAddresses) =>
-            throw new NotSupportedException();
+        public static unsafe SocketError TryGetAddrInfo(string name, bool justAddresses, out string? hostName, out string[] aliases, out IPAddress[] addresses, out int nativeErrorCode)
+        {
+            if (name == "")
+            {
+                // To match documented behavior on Windows, if an empty string is passed in, use the local host's name.
+                name = Dns.GetHostName();
+            }
+
+            Interop.Sys.HostEntry entry;
+            int result = Interop.Sys.GetHostEntryForName(name, &entry);
+            if (result != 0)
+            {
+                nativeErrorCode = result;
+                hostName = name;
+                aliases = Array.Empty<string>();
+                addresses = Array.Empty<IPAddress>();
+                return GetSocketErrorForNativeError(result);
+            }
+
+            ParseHostEntry(entry, justAddresses, out hostName, out aliases, out addresses);
+            nativeErrorCode = 0;
+            return SocketError.Success;
+        }
+
+        public static unsafe string? TryGetNameInfo(IPAddress addr, out SocketError socketError, out int nativeErrorCode)
+        {
+            byte* buffer = stackalloc byte[Interop.Sys.NI_MAXHOST + 1 /*for null*/];
+
+            byte isIPv6;
+            int rawAddressLength;
+            if (addr.AddressFamily == AddressFamily.InterNetwork)
+            {
+                isIPv6 = 0;
+                rawAddressLength = IPAddressParserStatics.IPv4AddressBytes;
+            }
+            else
+            {
+                isIPv6 = 1;
+                rawAddressLength = IPAddressParserStatics.IPv6AddressBytes;
+            }
+
+            byte* rawAddress = stackalloc byte[rawAddressLength];
+            addr.TryWriteBytes(new Span<byte>(rawAddress, rawAddressLength), out int bytesWritten);
+            Debug.Assert(bytesWritten == rawAddressLength);
+
+            int error = Interop.Sys.GetNameInfo(
+                rawAddress,
+                (uint)rawAddressLength,
+                isIPv6,
+                buffer,
+                Interop.Sys.NI_MAXHOST,
+                null,
+                0,
+                Interop.Sys.GetNameInfoFlags.NI_NAMEREQD);
+
+            socketError = GetSocketErrorForNativeError(error);
+            nativeErrorCode = error;
+            return socketError == SocketError.Success ? Marshal.PtrToStringAnsi((IntPtr)buffer) : null;
+        }
+
+        public static unsafe string GetHostName() => Interop.Sys.GetHostName();
+
+        public static unsafe Task GetAddrInfoAsync(string hostName, bool justAddresses)
+        {
```

Comment:
I see the case for `hostName == ""` is not handled. It's needed to be consistent to Windows.
Will add this plus a test therefore.